design/jira-traceability.md

```diff
@@ -0,0 +1,122 @@
+
+#### Jira Deployment UI and APIs:
+
+Jira has a separate section for Deployments in its work item UI. In this, only the highest deployed environment category is displayed. Eg. If the work item was deployed to a `test` type environment and  then to `production` type, Jira shows only `production` in deployment section. On clicking on this entry the user can see details of previous deployments.
+
+![Release1](./images/jira-traceability-deploy-1.png)
+
+Drill down view:
+![Release2](./images/jira-traceability-deploy-2.png)
+
+For the API, the combination of `pipelineId` (pipeline definition identifier), `environmentID` (environment identifier), `deploymentSequenceNumber` (deployment identifier) constitute a unique identifier for deployment. 
+
+
+#### When should the data be pushed from Azure Pipelines:?
+ Since build generation and deployments are done at a job level in azure pipelines, we'll push the status information to Jira at job boundaries. For deployments, azure pipelines already has a separate category of jobs, viz, deployment jobs. We'll invoke deployment APIs at the beginning and end of deployment jobs if all the filter conditions are met. For builds, however, we currently do not have a separate category of jobs. For this, we will assume that any job that publishes an artifact is a build job. We'll invoke build API at the end of these jobs if filter conditions are met.  
+
+
+
+## YAML: 
+
+Since traceability primarily involves interacting with external systems, we propose taking configuration for interaction with these external systems in a separate `integrations` section in YAML. As a beginning, we will support a first class integration with JIRA. In future, we may eventually make this section extensible thereby allowing 3rd party authors to create their own integrations. In this document, we will keep focus on JIRA integration. <em>Details on extensibility design are out of scope for this document.</em>
+
+ 
+
+### Schema:
+
+```yaml
+integrations:       
+  integration: jira # identifier of integration; mandatory;  
+  - connection: string # service connection to JIRA cloud; mandatory;
+    resources: # resource filters. commits only in  these resources will checked for linked JIRA issues. Job status will be posted to JIRA for these issues; optional; defaults to all resources
+    - resource: string # identifier for the resource;  
+      branches: 
+        include: [ string ] # optional; defaults to all branches.
+        exclude: [ string ] # optional; defaults to none
+    stages: [ string ]  # stage filters. Status of all build/deploy jobs in these stages will be posted to JIRA; optional; Defaults to all stages. 
+    environmentType: [ string ] # Environment type mapping as supported by JIRA; optional; By default all environments map to "unmapped" category;
```

Comment:
This should probably be called "environmentMap" or "environmentSpec". Also, it should be either a direct map of AzDO environment names to JIRA environment types _or_ a map of JIRA environment types to lists of AzDO environment names.

# Option 1
```yaml
# schema
environmentMap: { string: string }

# example
environmentMap:
  stage-US: staging
  East-US: production
  West-US: production
```

# Option 2
```yaml
# schema
environmentMap: { string: [ string ] }

# example
environmentMap:
  staging:
  - stage-US
  production:
  - East-US
  - West-US
```

Reply:
Makes sense. I have changed the schema to take environment details along similar lines.